### PR TITLE
MTDSA-19408: Validator Improvements for individuals-charges-api

### PR DESCRIPTION
**Pull request by GazalNebHMRC**

### CODE REVIEW

**Review by SamuelJacobsHMRC — CHANGES_REQUESTED** [review]

File: test/api/controllers/ControllerSpecHateoasSupport.scala [hunk 1]
```diff
@@ -16,13 +16,14 @@
...
-import api.models.hateoas.Method.GET
+import api.hateoas
+import api.hateoas.Link
+import api.hateoas.Method._
 import play.api.libs.json.{JsObject, Json}
 
 trait ControllerSpecHateoasSupport {
 
-  val hateoaslinks: Seq[Link] = Seq(Link(href = "/foo/bar", method = GET, rel = "test-relationship"))
+  val hateoaslinks: Seq[Link] = Seq(hateoas.Link(href = "/foo/bar", method = GET, rel = "test-relationship"))
```
[the remark this answers is not in the document]
Comment (GazalNebHMRC) [reply test/api/controllers/ControllerSpecHateoasSupport.scala:26]: Making it private is causing compilation errors.

File: .sbtopts [hunk 2]
```diff
@@ -0,0 +1,2 @@
+-J-Xmx3G
```
[the remark this answers is not in the document]
Comment (SamuelJacobsHMRC) [reply .sbtopts:1]: This file still needs removing

File: test/v2/controllers/validators/RetrievePensionChargesValidatorFactorySpec.scala [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    "return the parsed domain object" when {
+      "passed a valid request" in {
+        val result: Either[ErrorWrapper, RetrievePensionChargesRequestData] =
+          validator(validNino, validTaxYear).validateAndWrapResult()
+
+        result shouldBe Right(RetrievePensionChargesRequestData(parsedNino, parsedTaxYear))
+      }
+    }
+
+    // invalid nino
```
Comment (SamuelJacobsHMRC) [inline test/v2/controllers/validators/RetrievePensionChargesValidatorFactorySpec.scala:50]: comments in this file can be removed

File: app/v1/controllers/validators/AmendPensionChargesRulesValidator.scala [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+import api.models.errors._
+import cats.data.Validated
+import cats.data.Validated._
+import cats.implicits._
+import v1.models.request.AmendPensionCharges.{AmendPensionChargesRequestData, OverseasSchemeProvider, PensionCharges, PensionSavingsTaxCharges}
+
+object AmendPensionChargesRulesValidator extends RulesValidator[AmendPensionChargesRequestData] {
+  private val resolveParsedNumber      = ResolveParsedNumber()
+  private val qropsRefRegex            = "^[Q]{1}[0-9]{6}$"
+  private val pensionSchemeTaxRefRegex = "^\\d{8}[R]{1}[a-zA-Z]{1}$"
```
Comment (SamuelJacobsHMRC) [inline app/v1/controllers/validators/AmendPensionChargesRulesValidator.scala:30]:
```suggestion
  private val qropsRefRegex            = "^[Q]{1}[0-9]{6}$".r
  private val pensionSchemeTaxRefRegex = "^\\d{8}[R]{1}[a-zA-Z]{1}$".r
```

(to convert from string to regex)

File: app/v1/controllers/validators/AmendPensionChargesRulesValidator.scala [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+      .getOrElse(valid)
+  }
+
+  private def validateCharges(pensionCharges: PensionCharges): Validated[Seq[MtdError], Unit] = {
+    import pensionCharges._
+
+    val fieldsWithPaths = List(
+      (
+        pensionSavingsTaxCharges.flatMap(_.benefitInExcessOfLifetimeAllowance.map(_.amount)),
+        s"/pensionSavingsTaxCharges/benefitInExcessOfLifetimeAllowance/amount"),
```
Comment (SamuelJacobsHMRC) [inline app/v1/controllers/validators/AmendPensionChargesRulesValidator.scala:211]: lot's of strings starting with `s"..."` but no string interpolation being done so can be removed

**Review by SamuelJacobsHMRC — CHANGES_REQUESTED** [review]

File: .sbtopts (hunk 2 above)
[the remark this answers is not in the document]
Comment (GazalNebHMRC) [reply .sbtopts:1]: I added this to gitignore

**Review by SamuelJacobsHMRC — APPROVED** [review]

**Pull request merged**